File: src/EditorFeatures/CSharpTest/Structure/TypeDeclarationStructureTests.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CSharp.Structure;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Structure;
using Microsoft.CodeAnalysis.Test.Utilities;
using Roslyn.Test.Utilities;
using Xunit;

namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Structure;

[Trait(Traits.Feature, Traits.Features.Outlining)]
public sealed class TypeDeclarationStructureTests : AbstractCSharpSyntaxNodeStructureTests<TypeDeclarationSyntax>
{
    internal override AbstractSyntaxStructureProvider CreateProvider() => new TypeDeclarationStructureProvider();

    [Fact]
    public Task TestClass1()
        => VerifyBlockSpansAsync("""
                {|hint:$$class C{|textspan:
                {
                }|}|}
                """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Theory]
    [InlineData("enum")]
    [InlineData("class")]
    [InlineData("record")]
    [InlineData("record class")]
    [InlineData("record struct")]
    [InlineData("struct")]
    [InlineData("interface")]
    public Task TestClass2(string typeKind)
        => VerifyBlockSpansAsync($$"""
            {|hint:$$class C{|textspan:
            {
            }|}|}
            {{typeKind}}D
            {
            }
            """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Theory]
    [InlineData("enum")]
    [InlineData("class")]
    [InlineData("record")]
    [InlineData("record class")]
    [InlineData("record struct")]
    [InlineData("struct")]
    [InlineData("interface")]
    public Task TestClass3(string typeKind)
        => VerifyBlockSpansAsync($$"""
            {|hint:$$class C{|textspan:
            {
            }|}|}

            {{typeKind}}D
            {
            }
            """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact]
    public Task TestClassWithLeadingComments()
        => VerifyBlockSpansAsync("""
                {|span1:// Goo
                // Bar|}
                {|hint2:$$class C{|textspan2:
                {
                }|}|}
                """,
            Region("span1", "// Goo ...", autoCollapse: true),
            Region("textspan2", "hint2", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact]
    public Task TestClassWithNestedComments()
        => VerifyBlockSpansAsync("""
                {|hint1:$$class C{|textspan1:
                {
                    {|span2:// Goo
                    // Bar|}
                }|}|}
                """,
            Region("textspan1", "hint1", CSharpStructureHelpers.Ellipsis, autoCollapse: false),
            Region("span2", "// Goo ...", autoCollapse: true));

    [Fact]
    public Task TestInterface1()
        => VerifyBlockSpansAsync("""
                {|hint:$$interface I{|textspan:
                {
                }|}|}
                """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Theory]
    [InlineData("enum")]
    [InlineData("class")]
    [InlineData("record")]
    [InlineData("record class")]
    [InlineData("record struct")]
    [InlineData("struct")]
    [InlineData("interface")]
    public Task TestInterface2(string typeKind)
        => VerifyBlockSpansAsync($$"""
            {|hint:$$interface I{|textspan:
            {
            }|}|}
            {{typeKind}}D
            {
            }
            """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Theory]
    [InlineData("enum")]
    [InlineData("class")]
    [InlineData("record")]
    [InlineData("record class")]
    [InlineData("record struct")]
    [InlineData("struct")]
    [InlineData("interface")]
    public Task TestInterface3(string typeKind)
        => VerifyBlockSpansAsync($$"""
            {|hint:$$interface I{|textspan:
            {
            }|}|}

            {{typeKind}}D
            {
            }
            """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact]
    public Task TestInterfaceWithLeadingComments()
        => VerifyBlockSpansAsync("""
                {|span1:// Goo
                // Bar|}
                {|hint2:$$interface I{|textspan2:
                {
                }|}|}
                """,
            Region("span1", "// Goo ...", autoCollapse: true),
            Region("textspan2", "hint2", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact]
    public Task TestInterfaceWithNestedComments()
        => VerifyBlockSpansAsync("""
                {|hint1:$$interface I{|textspan1:
                {
                    {|span2:// Goo
                    // Bar|}
                }|}|}
                """,
            Region("textspan1", "hint1", CSharpStructureHelpers.Ellipsis, autoCollapse: false),
            Region("span2", "// Goo ...", autoCollapse: true));

    [Fact]
    public Task TestStruct1()
        => VerifyBlockSpansAsync("""
                {|hint:$$struct S{|textspan:
                {
                }|}|}
                """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Theory]
    [InlineData("enum")]
    [InlineData("class")]
    [InlineData("record")]
    [InlineData("record class")]
    [InlineData("record struct")]
    [InlineData("struct")]
    [InlineData("interface")]
    public Task TestStruct2(string typeKind)
        => VerifyBlockSpansAsync($$"""
            {|hint:$$struct C{|textspan:
            {
            }|}|}
            {{typeKind}}D
            {
            }
            """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Theory]
    [InlineData("enum")]
    [InlineData("class")]
    [InlineData("record")]
    [InlineData("record class")]
    [InlineData("record struct")]
    [InlineData("struct")]
    [InlineData("interface")]
    public Task TestStruct3(string typeKind)
        => VerifyBlockSpansAsync($$"""
            {|hint:$$struct C{|textspan:
            {
            }|}|}

            {{typeKind}}D
            {
            }
            """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact]
    public Task TestStructWithLeadingComments()
        => VerifyBlockSpansAsync("""
                {|span1:// Goo
                // Bar|}
                {|hint2:$$struct S{|textspan2:
                {
                }|}|}
                """,
            Region("span1", "// Goo ...", autoCollapse: true),
            Region("textspan2", "hint2", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact]
    public Task TestStructWithNestedComments()
        => VerifyBlockSpansAsync("""
                {|hint1:$$struct S{|textspan1:
                {
                    {|span2:// Goo
                    // Bar|}
                }|}|}
                """,
            Region("textspan1", "hint1", CSharpStructureHelpers.Ellipsis, autoCollapse: false),
            Region("span2", "// Goo ...", autoCollapse: true));

    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/62506")]
    public Task TestTypeDeclarationNoBraces1()
        => VerifyBlockSpansAsync("""
                {|comment:// comment|}
                $$struct S
                """,
            Region("comment", "// comment ...", autoCollapse: true));

    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/62506")]
    public Task TestTypeDeclarationNoBraces2()
        => VerifyBlockSpansAsync("""
                {|comment:// comment|}
                {|hint1:$$struct S{|textspan1:;|}|}
                """,
            Region("comment", "// comment ...", autoCollapse: true),
            Region("textspan1", "hint1", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/81662")]
    public Task TestExtension1()
        => VerifyBlockSpansAsync("""
                {|hint:$$extension{|textspan:(string s)
                {
                }|}|}
                """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));

    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/81662")]
    public Task TestExtension2()
        => VerifyBlockSpansAsync("""
                {|hint:$$extension<T>{|textspan:(string s)
                {
                }|}|}
                """,
            Region("textspan", "hint", CSharpStructureHelpers.Ellipsis, autoCollapse: false));
}

File: src/EditorFeatures/VisualBasicTest/Structure/TypeDeclarationStructureTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports Microsoft.CodeAnalysis.Structure
Imports Microsoft.CodeAnalysis.VisualBasic.Structure
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.Outlining
    <Trait(Traits.Feature, Traits.Features.Outlining)>
    Public Class TypeDeclarationStructureProviderTests
        Inherits AbstractVisualBasicSyntaxNodeStructureProviderTests(Of TypeStatementSyntax)

        Friend Overrides Function CreateProvider() As AbstractSyntaxStructureProvider
            Return New TypeDeclarationStructureProvider()
        End Function

        <Fact>
        Public Async Function TestClass() As Task
            Const code = "
{|span:Class $$C1
End Class|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "Class C1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestFriendClass() As Task
            Const code = "
{|span:Friend Class $$C1
End Class|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "Friend Class C1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestClassWithLeadingComments() As Task
            Const code = "
{|span1:'Hello
'World|}
{|span2:Class $$C1
End Class|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "' Hello ...", autoCollapse:=True),
                Region("span2", "Class C1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestClassWithNestedComments() As Task
            Const code = "
{|span1:Class $$C1
{|span2:'Hello
'World|}
End Class|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "Class C1 ...", autoCollapse:=False),
                Region("span2", "' Hello ...", autoCollapse:=True))
        End Function

        <Fact>
        Public Async Function TestModule() As Task
            Const code = "
{|span:Module $$M1
End Module|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "Module M1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestModuleWithLeadingComments() As Task
            Const code = "
{|span1:'Hello
'World|}
{|span2:Module $$M1
End Module|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "' Hello ...", autoCollapse:=True),
                Region("span2", "Module M1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestModuleWithNestedComments() As Task
            Const code = "
{|span1:Module $$M1
{|span2:'Hello
'World|}
End Module|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "Module M1 ...", autoCollapse:=False),
                Region("span2", "' Hello ...", autoCollapse:=True))
        End Function

        <Fact>
        Public Async Function TestInterface() As Task
            Const code = "
{|span:Interface $$I1
End Interface|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "Interface I1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestInterfaceWithLeadingComments() As Task
            Const code = "
{|span1:'Hello
'World|}
{|span2:Interface $$I1
End Interface|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "' Hello ...", autoCollapse:=True),
                Region("span2", "Interface I1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestInterfaceWithNestedComments() As Task
            Const code = "
{|span1:Interface $$I1
{|span2:'Hello
'World|}
End Interface|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "Interface I1 ...", autoCollapse:=False),
                Region("span2", "' Hello ...", autoCollapse:=True))
        End Function

        <Fact>
        Public Async Function TestStructure() As Task
            Const code = "
{|span:Structure $$S1
End Structure|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span", "Structure S1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestStructureWithLeadingComments() As Task
            Const code = "
{|span1:'Hello
'World|}
{|span2:Structure $$S1
End Structure|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "' Hello ...", autoCollapse:=True),
                Region("span2", "Structure S1 ...", autoCollapse:=False))
        End Function

        <Fact>
        Public Async Function TestStructureWithNestedComments() As Task
            Const code = "
{|span1:Structure $$S1
{|span2:'Hello
'World|}
End Structure|}
"

            Await VerifyBlockSpansAsync(code,
                Region("span1", "Structure S1 ...", autoCollapse:=False),
                Region("span2", "' Hello ...", autoCollapse:=True))
        End Function
    End Class
End Namespace
